astropy/io/fits/header.py

```diff
@@ -1694,11 +1694,11 @@ def _keyword_from_index(self, idx):
         """
 
         if idx < 0:
-            idx += len(self._cards) - 1
+            idx += len(self._cards)
 
         keyword = self._cards[idx].keyword
         keyword = Card.normalize_keyword(keyword)
-        repeat = self._keyword_indices[keyword].index(idx)
+        repeat = len(self._keyword_indices[keyword]) - 1
```

Comment:
I don't think this needs to be changed: though the docstring is unclear about `repeat`, from the usage of it in the `compressed` module it seems to be the sub-index of the keyword in the list of repeated keyword. At least it needs to be unique, which explains why with your change some tests are failing.